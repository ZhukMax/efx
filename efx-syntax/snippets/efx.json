{
  "Window": {
    "prefix": "Window",
    "body": [
      "<Window id=\"$1\" title=\"$2\" open=\"$3\" movable=\"$4\" resizable=\"$5\" default-width=\"$6\" default-height=\"$7\">",
      "  $0",
      "</Window>"
    ],
    "description": "EFx Window"
  },
  "Label": {
    "prefix": "Label",
    "body": [
      "<Label size=\"$1\" bold=\"$2\">$0</Label>"
    ],
    "description": "EFx Label"
  },
  "Button": {
    "prefix": "Button",
    "body": [
      "<Button rounding=\"$1\" enabled=\"$2\" tooltip=\"$3\">$0</Button>"
    ],
    "description": "EFx Button"
  },
  "Row": {
    "prefix": "Row",
    "body": [
      "<Row gap=\"$1\">",
      "  $0",
      "</Row>"
    ],
    "description": "EFx Row"
  },
  "Column": {
    "prefix": "Column",
    "body": [
      "<Column gap=\"$1\">",
      "  $0",
      "</Column>"
    ],
    "description": "EFx Column"
  },
  "Separator": {
    "prefix": "Separator",
    "body": [
      "<Separator/>"
    ],
    "description": "EFx Separator"
  },
  "Image (texture)": {
    "prefix": "Image",
    "body": [
      "<Image texture=\"$1\" width=\"$2\" height=\"$3\" rounding=\"$4\" clickable=\"$5\" tooltip=\"$6\"/>"
    ],
    "description": "EFx Image from texture/handle"
  },
  "Image (src)": {
    "prefix": "ImageSrc",
    "body": [
      "<Image src=\"$1\" max-width=\"$2\" maintain-aspect=\"$3\" id=\"$4\"/>"
    ],
    "description": "EFx Image from URI/path"
  },
  "Resize": {
    "prefix": "Resize",
    "body": [
      "<Resize id=\"$1\" default-height=\"$2\" min-height=\"$3\" max-height=\"$4\">",
      "  $0",
      "</Resize>"
    ],
    "description": "EFx Resize container"
  },
  "TextField": {
    "prefix": "TextField",
    "body": [
      "<TextField value=\"$1\" hint=\"$2\"/>"
    ],
    "description": "EFx TextField"
  },
  "TextField (password)": {
    "prefix": "Password",
    "body": [
      "<TextField value=\"$1\" hint=\"$2\" password=\"true\"/>"
    ],
    "description": "EFx TextField as password"
  },
  "Hyperlink": {
    "prefix": "Hyperlink",
    "body": [
      "<Hyperlink url=\"$1\" open_external=\"$2\" color=\"$3\" tooltip=\"$4\">$0</Hyperlink>"
    ],
    "description": "EFx Hyperlink with text"
  },
  "Hyperlink (self-closing)": {
    "prefix": "HyperlinkSelf",
    "body": [
      "<Hyperlink url=\"$1\" open_external=\"$2\" color=\"$3\" tooltip=\"$4\"/>"
    ],
    "description": "EFx Hyperlink (no inner text)"
  },
  "ScrollArea": {
    "prefix": "ScrollArea",
    "body": [
      "<ScrollArea axis=\"$1\" max-height=\"$2\" always-show=\"$3\" id=\"$4\">",
      "  $0",
      "</ScrollArea>"
    ],
    "description": "EFx ScrollArea (vertical|horizontal|both)"
  },
  "TopPanel": {
    "prefix": "TopPanel",
    "body": [
      "<TopPanel id=\"$1\" default-height=\"$2\" resizable=\"$3\" fill=\"$4\" stroke-width=\"$5\" stroke-color=\"$6\">",
      "  $0",
      "</TopPanel>"
    ],
    "description": "EFx TopPanel"
  },
  "BottomPanel": {
    "prefix": "BottomPanel",
    "body": [
      "<BottomPanel id=\"$1\" default-height=\"$2\" resizable=\"$3\" fill=\"$4\" stroke-width=\"$5\" stroke-color=\"$6\">",
      "  $0",
      "</BottomPanel>"
    ],
    "description": "EFx BottomPanel"
  },
  "SidePanel": {
    "prefix": "SidePanel",
    "body": [
      "<SidePanel side=\"$1\" id=\"$2\" default-width=\"$3\" min-width=\"$4\" resizable=\"$5\" fill=\"$6\" stroke-width=\"$7\" stroke-color=\"$8\">",
      "  $0",
      "</SidePanel>"
    ],
    "description": "EFx SidePanel (side=left|right)"
  },
  "CentralPanel": {
    "prefix": "CentralPanel",
    "body": [
      "<CentralPanel fill=\"$1\" padding=\"$2\" stroke-width=\"$3\" stroke-color=\"$4\">",
      "  $0",
      "</CentralPanel>"
    ],
    "description": "EFx CentralPanel"
  },
  "Panel": {
    "prefix": "Panel",
    "body": [
      "<Panel fill=\"$1\" padding=\"$2\" stroke-width=\"$3\" stroke-color=\"$4\" id=\"$5\">",
      "  $0",
      "</Panel>"
    ],
    "description": "EFx Panel (generic frame container)"
  },
  "Tabs": {
    "prefix": "Tabs",
    "body": [
      "<Tabs active=\"$1\" gap=\"$2\">",
      "  <Tab id=\"$3\" title=\"$4\">",
      "    $0",
      "  </Tab>",
      "</Tabs>"
    ],
    "description": "EFx Tabs with one Tab"
  },
  "Tab": {
    "prefix": "Tab",
    "body": [
      "<Tab id=\"$1\" title=\"$2\">",
      "  $0",
      "</Tab>"
    ],
    "description": "EFx Tab (child for <Tabs>)"
  },
  "Table": {
    "prefix": "Table",
    "body": [
      "<Table columns=\"$1\" striped=\"$2\" spacing-x=\"$3\" spacing-y=\"$4\" cell-align=\"$5\" id=\"$6\">",
      "  <Tr>",
      "    <Td>$0</Td>",
      "  </Tr>",
      "</Table>"
    ],
    "description": "EFx Table (Grid-based)"
  },
  "Tr": {
    "prefix": "Tr",
    "body": [
      "<Tr>",
      "  <Td>$0</Td>",
      "</Tr>"
    ],
    "description": "EFx Table row"
  },
  "Td": {
    "prefix": "Td",
    "body": [
      "<Td>$0</Td>"
    ],
    "description": "EFx Table cell"
  },
  "DataTable": {
    "prefix": "DataTable",
    "body": [
      "<DataTable id=\"$1\" striped=\"$2\" resizable=\"$3\" default-resizable=\"$4\" header-height=\"$5\" row-height=\"$6\" cell-align=\"$7\" cell-padding=\"$8\">",
      "  <Columns>",
      "    <Column mode=\"initial\" width=\"160\" resizable=\"true\"/>",
      "    <Column mode=\"auto\"/>",
      "    <Column mode=\"remainder\" resizable=\"true\" clip=\"true\"/>",
      "  </Columns>",
      "",
      "  <Header>",
      "    <Td><Label bold=\"true\">$9</Label></Td>",
      "    <Td><Label bold=\"true\">$10</Label></Td>",
      "    <Td><Label bold=\"true\">$11</Label></Td>",
      "  </Header>",
      "",
      "  <Tr>",
      "    <Td>$0</Td>",
      "    <Td></Td>",
      "    <Td></Td>",
      "  </Tr>",
      "</DataTable>"
    ],
    "description": "EFx DataTable (egui_extras::TableBuilder)"
  },
  "DataTable Column": {
    "prefix": "DTColumn",
    "body": [
      "<Column mode=\"$1\" width=\"$2\" resizable=\"$3\" clip=\"$4\"/>"
    ],
    "description": "EFx DataTable <Column> (auto|initial|exact|remainder)"
  },
  "DataTable Row": {
    "prefix": "DTRow",
    "body": [
      "<Tr height=\"$1\">",
      "  <Td>$0</Td>",
      "</Tr>"
    ],
    "description": "EFx DataTable row"
  },
  "Grid": {
    "prefix": "Grid",
    "body": [
      "<Grid columns=\"$1\" spacing-x=\"$2\" spacing-y=\"$3\" cell-align=\"$4\" cell-padding=\"$5\" striped=\"$6\" id=\"$7\">",
      "  $0",
      "</Grid>"
    ],
    "description": "EFx Grid (egui::Grid-based layout)"
  },
  "GridBreak": {
    "prefix": "GridBreak",
    "body": [
      "<GridBreak/>"
    ],
    "description": "EFx Grid row break"
  }
}