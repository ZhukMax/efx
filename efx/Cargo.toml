[package]
name = "efx"
description = "Rust XML templating engine for building egui UIs with procedural macros"
authors = ["Max Zhuk <mail@zhukmax.com>"]
version = "0.4.2"
edition = "2024"
rust-version = "1.85"
license = "MIT OR Apache-2.0"
readme = "../README.md"
repository = "https://github.com/ZhukMax/efx"
documentation = "https://docs.rs/efx"
keywords = ["egui", "eframe", "gui", "proc-macro", "templating"]
categories = [
    "gui",
    "template-engine",
    "development-tools::procedural-macro-helpers",
    "wasm",
]

[package.metadata.docs.rs]
# collect all the documentation with enabled features and cfg(docsrs) notes
all-features = true
rustdoc-args = ["--cfg", "docsrs"]

[badges]
maintenance = { status = "actively-developed" }

[lib]
proc-macro = true

[dependencies]
syn = { version = "2.0", features = ["full"] }
quote = "1.0"
proc-macro2 = "1.0"
efx-core = "1.0"

[dev-dependencies]
trybuild = "1.0"
# For the eframe example
eframe = "0.32"
# For the raw egui skeleton
egui = "0.32"
# For the bevy_egui example
bevy = { version = "0.16", default-features = false, features = [
    "bevy_winit",
    "bevy_render",
    "bevy_ui",
    "png",
    "x11"
] }
bevy_egui = { version = "0.36", default-features = false }
# for wasm entrypoint
wasm-bindgen = "0.2"
wasm-bindgen-futures = "0.4"
web-sys = { version = "0.3", features = ["HtmlCanvasElement", "Document", "Window"] }
