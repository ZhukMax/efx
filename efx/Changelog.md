## Changelog

#### 0.5
- Attribute rendering (efx-core)
- Label, Button, Row, Column, Separator attributes
- Updated documentation (attributes, quickstarts for eframe/bevy/winit+wgpu)
- New Tags: Heading, Hyperlink, TextField, Image, Grid
- Added Panel Tags: Window, CentralPanel, SidePanel, ScrollArea
- Added examples & tests

#### 0.4
- Basic Nodes Parser (efx-core) refactoring
- Added tag attributes

#### 0.3
- Divided the library into two crates
- Created Basic Nodes Parser
- Rewrote the macro to use the parser
- Tags: Label/Button/Row/Column/Separator
- Added tests

#### 0.2
- Refactoring expressions
- Added Button tag

#### 0.1
- Init engine
- Added rsx macros
- Added Label tag
