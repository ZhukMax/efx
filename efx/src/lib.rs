#![doc = include_str!("../docs/intro.md")]
#![doc = "\n\n---\n\n"]
#![doc = include_str!("../docs/tags.md")]
#![doc = "\n\n---\n\n"]
#![doc = include_str!("../docs/guide.md")]

mod attr_adapters;
mod input;
mod interfaces;
mod render;
mod tags;
mod utils;

use proc_macro::TokenStream;
use quote::quote;
use syn::parse_macro_input;

use crate::input::EfxInput;
use crate::render::render_nodes_as_stmts;
use crate::tags::Button;
use efx_core::{parse_str, Node};

/// Functional procedural macro `efx!` - parses compact XML-like markup
/// and executes it against the passed UI context.
///
/// Takes two arguments:
/// 1) **ui** — UI context expression/identifier;
/// 2) **template** — a string literal with markup.
/// # Example
/// ```rust
/// use efx_core::doc_prelude::*;
/// use efx::*;
///
/// let mut ui = Ui::default();
/// efx!(ui, r#"
///   <Column>
///      <Label>Hello</Label>
///      <Separator/>
///      <Row><Label>Row</Label></Row>
///   </Column>
/// "#);
/// ```
///
/// # Errors
/// - Unknown tag → `compile_error!`.
/// - `Separator` with children → `compile_error!`.
/// - Invalid interpolation `{ expr }` → `compile_error!`.
///
/// Tag attributes are **parsed** (since 0.4), but are currently **ignored** by the renderer.
#[proc_macro]
pub fn efx(input: TokenStream) -> TokenStream {
    let input = parse_macro_input!(input as EfxInput);
    let ui = input.ui;
    let template = input.template.value();

    let ast = match parse_str(&template) {
        Ok(nodes) => nodes,
        Err(err) => {
            let msg = format!("efx parse error: {}", err);
            return quote! { compile_error!(#msg); }.into();
        }
    };

    let expanded = if ast.len() == 1 {
        if let Node::Element(el) = &ast[0] {
            if el.name == "Button" {
                return render::render_tag::<Button>(&ui, el).into();
            } else {
                // Any other unit root is like a block with statements (return ())
                let body = render_nodes_as_stmts(&ui, &ast);
                quote! {{
                    #body
                }}
            }
        } else {
            // Text/interpolation on the root - just label
            let body = render_nodes_as_stmts(&ui, &ast);
            quote! {{
                #body
            }}
        }
    } else {
        let body = render_nodes_as_stmts(&ui, &ast);
        quote! {{
            #body
        }}
    };

    expanded.into()
}

#[proc_macro]
pub fn efx_ctx(input: TokenStream) -> TokenStream {
    use crate::input::EfxCtxInput;
    use efx_core::{parse_str, Node};

    let args = syn::parse_macro_input!(input as EfxCtxInput);
    let ctx_expr = args.ctx;
    let template = args.template.value();

    let ast = match parse_str(&template) {
        Ok(nodes) => nodes,
        Err(err) => {
            let msg = format!("efx parse error: {}", err);
            return quote! { compile_error!(#msg); }.into();
        }
    };

    let roots: Vec<Node> = ast
        .into_iter()
        .filter(|n| match n {
            Node::Text(t) => !t.value.trim().is_empty(),
            _ => true,
        })
        .collect();

    let root = match roots.as_slice() {
        // ровно один корневой элемент — OK
        [Node::Element(el)] => el.clone(),

        // пусто — ожидаем единственный корневой элемент
        [] => {
            let msg = "efx_ctx!: expected a single root element";
            return quote! { compile_error!(#msg); }.into();
        }

        // один узел, но не Element (например, текст/интерполяция) — ошибка
        [_] => {
            let msg = "efx_ctx!: root must be an element";
            return quote! { compile_error!(#msg); }.into();
        }

        // больше одного корневого узла — тоже ошибка
        [_, ..] => {
            let msg = "efx_ctx!: expected a single root element";
            return quote! { compile_error!(#msg); }.into();
        }
    };

    let allowed = [
        "Window",
        "CentralPanel",
        "TopPanel",
        "BottomPanel",
        "SidePanel",
    ];
    if !allowed.iter().any(|&n| n == root.name) {
        let msg = format!(
            "efx_ctx!: root <{}> is not context-root. Use efx!(ui, ...) instead.",
            root.name
        );
        return quote! { compile_error!(#msg); }.into();
    }

    // Generating a render for the root
    let ui_param = quote!(__efx_ui_ctx);
    let body = render::render_element_stmt(&ui_param, &root);

    quote! {{
        // local wrapper: send only .ctx()
        struct __EfxUiCtxShim<'a>(&'a egui::Context);
        impl<'a> __EfxUiCtxShim<'a> {
            #[inline]
            fn ctx(&self) -> &egui::Context { self.0 }
        }

        // clone the context and render the root
        let __efx_ctx_local = (#ctx_expr).clone();
        let __efx_ui_ctx = __EfxUiCtxShim(&__efx_ctx_local);
        #body
        ()
    }}
    .into()
}
