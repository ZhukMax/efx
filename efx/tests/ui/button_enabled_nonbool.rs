use efx::efx;

#[derive(Default)]
struct Ui;
impl Ui {
    fn button<S: Into<String>>(&mut self, _s: S) -> Resp { Resp }
    fn add<T>(&mut self, _w: T) -> Resp { Resp }
    fn add_enabled<T>(&mut self, _b: bool, _w: T) -> Resp { Resp }
}
struct Resp; impl Resp { fn clicked(&self) -> bool { false } }

fn main() {
    let mut ui = Ui::default();
    // enabled must be true/false
    let _ = efx!(ui, r#"<Button enabled="maybe">Run</Button>"#);
}
