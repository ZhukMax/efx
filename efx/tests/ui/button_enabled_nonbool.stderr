error: macro expansion ignores `;` and any tokens following
  --> tests/ui/button_enabled_nonbool.rs:15:13
   |
15 |     let _ = efx!(ui, r#"<Button enabled="maybe">Run</Button>"#);
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ caused by the macro expansion here
   |
   = note: the usage of `efx!` is likely invalid in expression context

error: efx: attribute `enabled` expects boolean (true/false), got `maybe`
  --> tests/ui/button_enabled_nonbool.rs:15:13
   |
15 |     let _ = efx!(ui, r#"<Button enabled="maybe">Run</Button>"#);
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `efx` (in Nightly builds, run with -Z macro-backtrace for more info)
